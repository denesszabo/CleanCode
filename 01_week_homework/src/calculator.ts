export class Calculator {

    public Add(a: number, b: number): number {
        return a + b;
    }

    public Sub(a: number, b: number): number {
        return a - b;
    }

}
