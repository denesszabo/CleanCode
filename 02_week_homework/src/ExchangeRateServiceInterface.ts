/**
 * Defines the currency exchange rate service.
 */
export interface ExchangeRateServiceInterface {

    /**
     * Get the exchange rate between two currencies.
     *
     * @param fromCurrency string
     *   The currency to convert from.
     * @param toCurrency string
     *   The currency to convert to.
     *
     * @trows InvalidCurrencyError
     *   If the given currency is invalid or unknown
     */
    getExchangeRate(fromCurrency: string, toCurrency: string): number;
}
