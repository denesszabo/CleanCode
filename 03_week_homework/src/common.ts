export function delay(millisecond: number) : Promise<void> {
    return new Promise<void>((resolve) => {
        setTimeout(resolve, millisecond);
    })
}
