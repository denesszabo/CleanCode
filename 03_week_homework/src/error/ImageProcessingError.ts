export class ImageProcessingError extends Error {
    public constructor(message: string) {
        super(message);
        this.name = this.constructor.name;
    }
}
