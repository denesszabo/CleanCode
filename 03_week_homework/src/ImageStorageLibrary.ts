import {delay} from "./common";

export class ImageStorageLibrary {

    public async saveContentIntoFile(destination: string, content: string): Promise<void> {
        console.log(`Saving file content ${content} to path ${destination}`);

        // Simulate the actual image saving process. It takes a time.
        await delay(1000);
    }
}
