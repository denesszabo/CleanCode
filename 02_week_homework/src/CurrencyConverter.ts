import { ExchangeRateServiceInterface } from "./ExchangeRateServiceInterface";
import {InvalidInputAmountError} from "./error/InvalidInputAmountError";
import {InvalidCurrencyError} from "./error/InvalidCurrencyError";
import {NetworkError} from "./error/NetworkError";

export class CurrencyConverter {

    private readonly FIXED_AMOUNT = 100;
    private readonly EXCHANGE_RATE_ERROR = -1;

    constructor(private exchangeRateService: ExchangeRateServiceInterface) { }

    public Convert(amount: number, fromCurrency: string, toCurrency: string): number {
        this.validateAmount(amount);
        const exchangeRate = this.getExchangeRate(fromCurrency, toCurrency);

        this.validateExchangeRate(exchangeRate);
        return amount * exchangeRate;
    }

    /**
     * Conversion report builder for the give date range.
     *
     * @param fromCurrency
     * @param toCurrency
     * @param startDate
     * @param endDate
     * @constructor
     */
    public GenerateConversionReport(
        fromCurrency: string,
        toCurrency: string,
        startDate: Date,
        endDate: Date
    ): string {
        const conversions: number[] = [];

        const currentDate = new Date(startDate);

        while (currentDate <= endDate) {

            try {
                const exchangeRate: number | null = this.exchangeRateService.getExchangeRate(fromCurrency, toCurrency);
                if (exchangeRate != this.EXCHANGE_RATE_ERROR) {
                    this.validateExchangeRate(exchangeRate);
                    this.calculateConversion(exchangeRate, conversions, currentDate);
                }
            }
            catch (error) {
                // @todo: Inject a log service.
                // console.log(error.message, error);
                // return `Conversion Report:\n${conversions.join('\n')}`;
                break;
            }

        }

        return `Conversion Report:\n${conversions.join('\n')}`;
    }

    /**
     * Service call for exchange rate.
     *
     * @param fromCurrency
     *   The currency to convert from.
     * @param toCurrency
     *   The currency to convert to.
     *
     * @return number|null
     *   The exchange rate or null if an error occurred.
     *
     * @private
     */
    private getExchangeRate(fromCurrency: string, toCurrency: string)    {
        try {
            return this.exchangeRateService.getExchangeRate(fromCurrency, toCurrency);
        }
        catch (error) {
            if (error instanceof InvalidCurrencyError) {
                // @todo: Inject a log service.
                console.log(error.message, error);
                return this.EXCHANGE_RATE_ERROR;
            }
            console.log(error);
            throw new NetworkError('The network is down.');
        }

    }

    private calculateConversion(exchangeRate: number, conversions: number[], currentDate: Date) {
        const convertedAmount = this.FIXED_AMOUNT * exchangeRate; // Assume a fixed amount for simplicity
        conversions.push(convertedAmount);
        currentDate.setDate(currentDate.getDate() + 1);
    }

    private validateExchangeRate(exchangeRate: number) {
        if (!exchangeRate) {
            throw new Error('Unable to fetch exchange rate.');
        }

        if (isNaN(exchangeRate)) {
            throw new Error('Invalid exchange rate.');
        }
    }

    /**
     * Validate the input amount.
     *
     * @param amount number
     *   The amount to validate.
     *
     * @throws InvalidInputAmountError
     *   If the amount is invalid.
     * @throws RangeError
     *   If the amount is negative.
     *
     * @private
     */
    private validateAmount(amount: number): void {
        if (isNaN(amount)) {
            throw new InvalidInputAmountError('Invalid amount input.');
        }
        if (amount < 0) {
            throw new RangeError('The amount must be a positive number.');
        }
    }
}
