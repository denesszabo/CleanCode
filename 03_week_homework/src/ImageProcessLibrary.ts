import { delay } from "./common";
import { InvalidImageError } from "./error/InvalidImageError";
import { ImageProcessingError } from "./error/ImageProcessingError";

export class ImageProcessLibrary {
    public static readonly SUCCESS = 1;
    protected processedImage: string;

    public constructor() {
        // Set the processed image to an empty string.
        this.processedImage = '';
    }

    /**
     * Process an image.
     *
     * @param source
     *   The source image.
     * @param destination
     *   The destination path.
     *
     * @throws InvalidImageError.
     * @throws ImageProcessingError
     */
    public async processImage(source: string, destination: string): Promise<number> {
        this.processedImage = '';

        try {
            this.validateImage(source);
            this.validateImage(destination);
            console.log(`Processing image from ${source} to ${destination}`);

            // Simulate the actual image processing process. It takes a time.
            await delay(1500);

            console.log('Image processed successfully.');

            // @todo: Should throw an error if the image cannot be processed.
            // @todo: Implement the image processing here.
            this.processedImage = 'PROCESSED_IMAGE_CONTENT';
            return ImageProcessLibrary.SUCCESS;
        }
        catch (e) {
            let error = e as Error;
            console.error(error.message);
            throw error;
        }

    }

    protected validateImage(image: string) {
        if (image.length == 0) {
            throw new InvalidImageError('The image cannot be empty.');
        }
    }


    public getProcessedImage(): string {
        return this.processedImage;
    }
}
