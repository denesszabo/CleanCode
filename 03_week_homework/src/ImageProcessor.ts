import { ImageProcessLibrary } from "./ImageProcessLibrary";
import { ImageStorageLibrary } from "./ImageStorageLibrary";
import { InvalidImageError } from "./error/InvalidImageError";

export class ImageProcessor {
    public static readonly SUCCESS = 1;
    public static readonly FAILED = -1;

    public constructor(protected imageProcessLibrary: ImageProcessLibrary, protected imageStorageLibrary: ImageStorageLibrary) {
    }

    public async processAndSaveImage(source: string, destination: string): Promise<number> {

        try {
            console.log(`Pocessing image ${source} to ${destination}.`);
            this.validateImage(source);
            this.imageProcessLibrary.processImage(source, destination);

            const processedImageContent: string = this.imageProcessLibrary.getProcessedImage();
            await this.imageStorageLibrary.saveContentIntoFile(destination, processedImageContent);
        }
        catch (e) {
            let error = e as Error;
            // console.error(error.message);
            console.log(error.message);
            throw error;
        }
        // Should return the processed image content.
        return ImageProcessor.SUCCESS;
    }

    public validateImage(image: string): void    {
        if (!image.endsWith('.jpg')) {
            throw new InvalidImageError('Invalid image format. Only JPG images are supported.');
        }
    }

}
