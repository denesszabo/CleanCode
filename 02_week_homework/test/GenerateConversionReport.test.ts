import {CurrencyConverter} from "../src/CurrencyConverter";

import {mock, mockReset} from "jest-mock-extended";
import {ExchangeRateServiceInterface} from "../src/ExchangeRateServiceInterface";
import {InvalidCurrencyError} from "../src/error/InvalidCurrencyError";
import {NetworkError} from "../src/error/NetworkError";
const mockedExchangeRateService = mock<ExchangeRateServiceInterface>();

describe('Generate Conversion report tests', () => {
    let sut: CurrencyConverter;

    beforeAll(() => {
        mockReset(mockedExchangeRateService);
        sut = new CurrencyConverter(mockedExchangeRateService);
    })

    describe('happy part of the generations', () => {
        it('should generate conversion report', () => {
            // Arrange
            const conversionReport = new CurrencyConverter(mockedExchangeRateService);
            const from = 'HUF';
            const to = 'EUR';
            const startDate = new Date();
            const endDate = new Date();
            const days = 10;
            endDate.setDate(endDate.getDate() + days - 1);
            mockedExchangeRateService.getExchangeRate.calledWith('HUF', 'EUR').mockReturnValue(0.0026);

            // Act
            let report = conversionReport.GenerateConversionReport(from, to, startDate, endDate);
            // console.log(report);

            // Assert
            expect(report.length == days);
            expect(report).toMatchSnapshot();
            // This is a private meth, can't be spied. (I have not known yet how to spy private methods).
            // expect(sut.exchangeRateService).toBeCalledTimes(days);
            expect(mockedExchangeRateService.getExchangeRate).toBeCalledTimes(days);
        });
    })


    describe('Sad part of the generations', () => {
        it('should generate empty conversion report', () => {
            // Arrange
            const conversionReport = new CurrencyConverter(mockedExchangeRateService);
            const from = '';
            const to = 'EUR';
            const startDate = new Date();
            const endDate = new Date();
            const days = 10;
            endDate.setDate(endDate.getDate() + days - 1);
            const expectedError = new InvalidCurrencyError("The from currency is invalid.");

            // mockedExchangeRateService.getExchangeRate.calledWith('HUF', 'EUR').mockReturnValue(0.0026);
            mockedExchangeRateService.getExchangeRate.mockImplementation((from, to) => {throw expectedError});

            // Act
            let report = conversionReport.GenerateConversionReport(from, to, startDate, endDate);

            // Assert
            expect(report.length == 0);
        });

        it('should generate empty conversion by network error', () => {
            // Arrange
            const conversionReport = new CurrencyConverter(mockedExchangeRateService);
            const from = 'HUF';
            const to = 'EUR';
            const startDate = new Date();
            const endDate = new Date();
            const days = 10;
            endDate.setDate(endDate.getDate() + days - 1);
            const expectedError = new NetworkError("The network is down.");

            // mockedExchangeRateService.getExchangeRate.calledWith('HUF', 'EUR').mockReturnValue(0.0026);
            mockedExchangeRateService.getExchangeRate.mockImplementation((from, to) => {throw expectedError});

            // Act
            let report = conversionReport.GenerateConversionReport(from, to, startDate, endDate);

            // Assert
            expect(report.length == 0);
        });
    })

})
