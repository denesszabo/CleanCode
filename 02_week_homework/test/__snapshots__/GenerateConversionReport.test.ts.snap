// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate Conversion report tests happy part of the generations should generate conversion report 1`] = `
"Conversion Report:
0.26
0.26
0.26
0.26
0.26
0.26
0.26
0.26
0.26
0.26"
`;
